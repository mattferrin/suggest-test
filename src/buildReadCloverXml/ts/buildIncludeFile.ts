import { MeowOutput } from "../../bin/ts/cli";
import { SummaryItem } from "./undefinedOrSummary";

export function buildIncludeFile(input: MeowOutput) {
  return function includeFile(file: SummaryItem | undefined) {
    const isTested =
      (file?.statements ?? 0) > (file?.uncoveredStatements ?? 0) ||
      (file?.conditionals ?? 0) > (file?.uncoveredConditionals ?? 0);
    const isIncludeTested = input.flags.includeTested && isTested;

    const isNotTested =
      (file?.statements ?? 0) === (file?.uncoveredStatements ?? 0) &&
      (file?.conditionals ?? 0) === (file?.uncoveredConditionals ?? 0);
    const isIncludeNotTested = input.flags.includeNotTested && isNotTested;

    return isIncludeTested || isIncludeNotTested;
  };
}
