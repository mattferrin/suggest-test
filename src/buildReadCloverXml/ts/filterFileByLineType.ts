import { MeowOutput } from "../../bin/ts/cli";
import { FileArray } from "./summarizedResult";
import { Untrusted } from "./Untrusted";

export function filterFileByLineType(
  input: MeowOutput,
  projectPackage: Untrusted<{
    readonly file: FileArray;
  }>
): Untrusted<FileArray> {
  if (!input.flags.includeStatements) {
    return projectPackage?.file?.filter((projectPackageFile) => {
      return projectPackageFile?.line?.some((line) => {
        return line?.$?.type === "cond";
      });
    });
  } else {
    return projectPackage?.file;
  }
}
