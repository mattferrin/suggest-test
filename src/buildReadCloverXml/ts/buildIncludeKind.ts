import { MeowOutput } from "../../bin/ts/cli";
import { SummaryItem } from "./undefinedOrSummary";

export function buildIncludeKind(input: MeowOutput) {
  return function includeKind(file: SummaryItem | undefined) {
    const isStatementKeeper =
      (file?.statements ?? 0) > 1 && (file?.uncoveredStatements ?? 0) > 0;
    const isConditionalKeeper =
      (file?.conditionals ?? 0) > 1 && (file?.uncoveredConditionals ?? 0) > 0;

    if (!input.flags.includeConditionals) {
      return isStatementKeeper;
    } else if (!input.flags.includeStatements) {
      return isConditionalKeeper;
    } else {
      return isStatementKeeper || isConditionalKeeper;
    }
  };
}
