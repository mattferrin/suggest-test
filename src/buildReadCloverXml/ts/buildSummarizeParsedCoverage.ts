import { MeowOutput } from "../../bin/ts/cli";
import { filterFileByLineType } from "./filterFileByLineType";
import { FileArray, summarizedResult } from "./summarizedResult";
import { SummaryItem } from "./undefinedOrSummary";
import { Untrusted } from "./Untrusted";

export type Parsed = Untrusted<{
  readonly coverage: {
    readonly project: ReadonlyArray<{
      readonly package: ReadonlyArray<{
        readonly file: FileArray;
      }>;
    }>;
  };
}>;

type SummarizeParsedCoverageOutput =
  | ReadonlyArray<SummaryItem | undefined>
  | undefined;

export function buildSummarizeParsedCoverage(input: MeowOutput) {
  return function summarizeParsedCoverage(
    parsed: Parsed
  ): SummarizeParsedCoverageOutput {
    return parsed?.coverage?.project?.flatMap((project) => {
      return project?.package?.flatMap((projectPackage) => {
        const filteredProjectPackageFile = filterFileByLineType(
          input,
          projectPackage
        );

        return summarizedResult(input, filteredProjectPackageFile);
      });
    });
  };
}
