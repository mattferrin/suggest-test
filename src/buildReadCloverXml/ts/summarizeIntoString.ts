/* eslint-disable functional/functional-parameters */
/* eslint-disable @typescript-eslint/restrict-template-expressions */

import { MeowOutput } from "../../bin/ts/cli";
import { SummaryItem } from "./undefinedOrSummary";

export function summarizeIntoString(
  input: MeowOutput,
  summaryArray: ReadonlyArray<SummaryItem | undefined>
): string {
  if (!input.flags.includeConditionals && !input.flags.includeStatements) {
    return "\nBoth includeConditionals and includeStatements are false. Please let at least 1 be true.\n";
  } else {
    const entity = [
      ...(input.flags.includeConditionals ? ["conditionals"] : []),
      ...(input.flags.includeStatements ? ["statements"] : []),
    ];

    if (summaryArray.length === 0) {
      return `\nNo uncovered ${entity.join(" or ")} were found.\n`;
    } else {
      const outputLines = summaryArray.map((line) => {
        return `${line?.path ?? "MISSING/PATH"} ${new Intl.NumberFormat(
          "en-US",
          {
            maximumSignificantDigits: 2,
          }
        ).format(line?.score ?? 0)}`;
      });

      return [
        `\nFiles suggested to increase coverage of ${entity.join(" and ")}:\n`,
        ...outputLines,
        "",
      ]
        .filter((line) => line !== null)
        .join("\n");
    }
  }
}
