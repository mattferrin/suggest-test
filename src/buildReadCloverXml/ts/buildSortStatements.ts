import { MeowOutput } from "../../bin/ts/cli";
import { ReadCloverXmlOutput } from "./buildReadCloverXml";
import { summarizeIntoString } from "./summarizeIntoString";

export function buildSortStatements(input: MeowOutput) {
  return function sortStatements(
    output: ReadCloverXmlOutput
  ): ReadCloverXmlOutput {
    switch (output.tag) {
      case "errors":
        return output;
      case "summary": {
        const summaryArray = output.summaryArray
          ?.filter((file) => (file?.statements ?? 0) > 1)
          ?.sort((fileA, fileB) => {
            if (
              fileB?.statements === null ||
              fileB?.statements === undefined ||
              fileA?.statements === null ||
              fileA?.statements === undefined
            ) {
              return -1;
            } else {
              return fileA.weightedStatements > fileB.weightedStatements
                ? -1
                : 1;
            }
          });

        return {
          tag: "summary",
          summaryArray,
          summaryString: summarizeIntoString(summaryArray),
        };
      }
    }
  };
}
