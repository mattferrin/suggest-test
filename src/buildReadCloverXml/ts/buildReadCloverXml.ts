import fs from "fs/promises";
import { parseStringPromise } from "xml2js";
import { MeowOutput } from "../../bin/ts/cli";
import { buildSortStatements } from "./buildSortStatements";
import { buildSummarizeParsedCoverage } from "./buildSummarizeParsedCoverage";
import { passThroughLogger } from "./passThroughLogger";
import { SummaryItem, undefinedOrSummary } from "./undefinedOrSummary";

export type ReadCloverXmlOutput =
  | { readonly tag: "errors"; readonly errors: readonly Error[] }
  | {
      readonly tag: "summary";
      readonly summaryArray: ReadonlyArray<SummaryItem | undefined>;
      readonly summaryString: string | { readonly tag: "not-ready" };
    };

export function buildReadCloverXml(input: MeowOutput) {
  // eslint-disable-next-line functional/functional-parameters
  return async function readCloverXml(): Promise<ReadCloverXmlOutput> {
    return await fs
      .readFile(input.input[0] ?? "coverage/clover.xml")
      .then(parseStringPromise)
      .then(buildSummarizeParsedCoverage(input))
      .then(undefinedOrSummary)
      .then(buildSortStatements(input))
      .catch((reason): ReadCloverXmlOutput => {
        return { tag: "errors", errors: [new Error(reason)] };
      })
      .then(passThroughLogger);
  };
}
