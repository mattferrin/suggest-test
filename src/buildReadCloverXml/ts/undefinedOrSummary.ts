import { ReadCloverXmlOutput } from "./buildReadCloverXml";

export interface SummaryItem {
  readonly conditionals: number | null | undefined;
  readonly path: string | null | undefined;
  readonly statements: number | null | undefined;
  readonly uncoveredConditionals: number;
  readonly uncoveredStatements: number;
  readonly score: number;
  readonly hasUnmarkedConditional: boolean;
}

export function undefinedOrSummary(
  summary: ReadonlyArray<SummaryItem | undefined> | undefined
): ReadCloverXmlOutput {
  if (summary === undefined || summary === null) {
    return {
      tag: "errors",
      errors: [
        new Error(
          "The parsed clover.xml has an undefined or null property somewhere"
        ),
      ],
    };
  } else {
    return {
      tag: "summary",
      summaryArray: summary,
      summaryString: { tag: "not-ready" },
    };
  }
}
