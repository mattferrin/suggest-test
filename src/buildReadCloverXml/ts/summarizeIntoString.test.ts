import { MeowOutput } from "../../bin/ts/cli";
import { summarizeIntoString } from "./summarizeIntoString";
import { SummaryItem } from "./undefinedOrSummary";

function snap(input: {
  stabileSortId: string;
  comment: string | null;
  input: MeowOutput;
  summaryArray: ReadonlyArray<SummaryItem | undefined>;
}): void {
  it(input.stabileSortId, () => {
    /** setup mocks */

    /** unit under test */
    const result = summarizeIntoString(input.input, input.summaryArray);

    /** assertions */
    expect({
      [`${input.comment ?? "TODO: comment"}`]: {
        result,
      },
    }).toMatchSnapshot();
  });
}

snap({
  stabileSortId: "63ae7c30-ed68-4726-b3a6-20deb68b73bd",
  comment: "show please let 1 include be true",
  input: {
    flags: { includeConditionals: false, includeStatements: false },
  } as MeowOutput,
  summaryArray: [],
});

snap({
  stabileSortId: "4337e511-5d22-4675-8ef1-adf84cc74172",
  comment: "show no statements",
  input: {
    flags: { includeConditionals: false, includeStatements: true },
  } as MeowOutput,
  summaryArray: [],
});
snap({
  stabileSortId: "51f963fb-6201-47b5-b164-d0b8dfb9b612",
  comment: "show no conditionals",
  input: {
    flags: { includeConditionals: true, includeStatements: false },
  } as MeowOutput,
  summaryArray: [],
});
snap({
  stabileSortId: "f91efabb-ab0a-4a8a-acf2-0829970bb814",
  comment: "show no conditionals or statements",
  input: {
    flags: { includeConditionals: true, includeStatements: true },
  } as MeowOutput,
  summaryArray: [],
});

snap({
  stabileSortId: "279dd56b-7501-46e8-a08d-a7588ac37114",
  comment: "missing path or score don't error",
  input: {
    flags: { includeConditionals: true, includeStatements: true },
  } as MeowOutput,
  summaryArray: [{}] as SummaryItem[],
});

snap({
  stabileSortId: "41fdcfa1-e922-4d60-8d20-67395195c7e6",
  comment: "shows 1 line",
  input: {
    flags: { includeConditionals: true, includeStatements: true },
  } as MeowOutput,
  summaryArray: [{ path: "actual/path", score: 99 } as SummaryItem],
});
snap({
  stabileSortId: "41fdcfa1-e922-4d60-8d20-67395195c7e6",
  comment: "shows 2 lines",
  input: {
    flags: { includeConditionals: true, includeStatements: true },
  } as MeowOutput,
  summaryArray: [
    { path: "actual/path", score: 99 },
    { path: "actual/path/1", score: 999 },
  ] as SummaryItem[],
});
