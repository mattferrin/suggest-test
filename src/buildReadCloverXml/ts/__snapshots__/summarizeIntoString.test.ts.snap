// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`41fdcfa1-e922-4d60-8d20-67395195c7e6 1`] = `
{
  "shows 1 line": {
    "result": "
Files suggested to increase coverage of conditionals and statements:

actual/path 99
",
  },
}
`;

exports[`41fdcfa1-e922-4d60-8d20-67395195c7e6 2`] = `
{
  "shows 2 lines": {
    "result": "
Files suggested to increase coverage of conditionals and statements:

actual/path 99
actual/path/1 1,000
",
  },
}
`;

exports[`51f963fb-6201-47b5-b164-d0b8dfb9b612 1`] = `
{
  "show no conditionals": {
    "result": "
No uncovered conditionals were found.
",
  },
}
`;

exports[`63ae7c30-ed68-4726-b3a6-20deb68b73bd 1`] = `
{
  "show please let 1 include be true": {
    "result": "
Both includeConditionals and includeStatements are false. Please let at least 1 be true.
",
  },
}
`;

exports[`279dd56b-7501-46e8-a08d-a7588ac37114 1`] = `
{
  "missing path or score don't error": {
    "result": "
Files suggested to increase coverage of conditionals and statements:

MISSING/PATH 0
",
  },
}
`;

exports[`4337e511-5d22-4675-8ef1-adf84cc74172 1`] = `
{
  "show no statements": {
    "result": "
No uncovered statements were found.
",
  },
}
`;

exports[`f91efabb-ab0a-4a8a-acf2-0829970bb814 1`] = `
{
  "show no conditionals or statements": {
    "result": "
No uncovered conditionals or statements were found.
",
  },
}
`;
